Zapper-fi/studio
# feat(pancakeswap-v3): Added basic integration for PancakeSwap V3

## Description

This PR features a basic integration of PancakeSwap V3, which contains a lot of the code from Uniswap V3. It required adding two new libraries to `package.json` 
- `@pancakeswap/v3-sdk`
- `@pancakeswap/swap-sdk-core`

They're both forks of the Uniswap V3 sdk / core libraries.

Most of the code reused **could likely be abstracted away instead of copied**. Wanted to open a draft PR to also discuss if it'd be interesting to create a Uniswap base and simply override a few methods when if comes to supporting its forks. This can get messy pretty quickly otherwise, especially considering there are quite a few forks of Uniswap.

I've also added all `abis` even though only 3 are being used. This will require a tiny clean up. 

Lastly, Pancakeswap V3 **did not have a Pool factory** like Uniswap does. The solution I could find was to log one of the pool's addresses from an address that transacted with the Pool Deployer to create a new Pool, and get the `abi` from that address while making sure it followed the `IPancakeV3Pool` interface. Their documentation did not include strategies to instantiate a `PancakeV3Pool` so I had to improvise.  

## Checklist


## How to test?

This will require adding `pancakeswap-v3` to the supports apps in the `.env` file like so : 

```
ENABLED_APPS=pancakeswap-v3
```

While running the studio in dev with `pnpm dev`, positions and balances can be queries using the regular URL endpoints. Make sure to use the `binance-smart-chain` network. I've been using a few addresses that made recent transactions on a PancakeSwap pool. Here's one : 0xBB905c145120C3b4a36b1a4F10F78B957842945f

### Positions
http://localhost:5001/apps/pancakeswap-v3/positions?groupIds[]=liquidity&network=binance-smart-chain

### Balance
http://localhost:5001/apps/pancakeswap-v3/balances?addresses[]=0xBB905c145120C3b4a36b1a4F10F78B957842945f&network=binance-smart-chain
